src/mumble/Overlay_win.cpp

```diff
@@ -97,26 +169,101 @@ static const char *exitStatusString(QProcess::ExitStatus exitStatus) {
+		case QProcess::WriteError:
+			return "an error occurred when attempting to write to the process";
+		case QProcess::ReadError:
+			return "an error occurred when attempting to read from the process";
+	}
+
+	return "unknown";
+}
+
 void OverlayPrivateWin::onHelperProcessStarted() {
 	QProcess *helper = qobject_cast<QProcess *>(sender());
+	QString path;
+	if (helper == m_helper_process) {
+		path = m_helper_exe_path;
+	} else if (helper == m_helper64_process) {
+		path = m_helper64_exe_path;
+	}
 
 	PROCESS_INFORMATION *pi = helper->pid();
 	qWarning("OverlayPrivateWin: overlay helper process '%s' started with PID %llu.",
-	         qPrintable(m_helper_exe_path), static_cast<unsigned long long>(pi->dwProcessId));
+	         qPrintable(path), static_cast<unsigned long long>(pi->dwProcessId));
+}
+
+void OverlayPrivateWin::onHelperProcessError(QProcess::ProcessError processError) {
+	QProcess *helper = qobject_cast<QProcess *>(sender());
+	QString path;
+	if (helper == m_helper_process) {
+		path = m_helper_exe_path;
+	} else if (helper == m_helper64_process) {
+		path = m_helper64_exe_path;
+	}
+
+	qWarning("OverlayPrivateWin: an error occured for overlay helper process '%s': %s",
+	         qPrintable(path),
+	         processErrorString(processError));
 }
 
 void OverlayPrivateWin::onHelperProcessExited(int exitCode, QProcess::ExitStatus exitStatus) {
 	QProcess *helper = qobject_cast<QProcess *>(sender());
 
+	qint64 elapsedMsec;
+	QString path;
+	if (helper == m_helper_process) {
+		path = m_helper_exe_path;
+		elapsedMsec = m_helper_start_time.elapsed();
+	} else if (helper == m_helper64_process) {
+		path = m_helper64_exe_path;
+		elapsedMsec = m_helper64_start_time.elapsed();
+	}
+
 	const char *helperErrString = OverlayHelperErrorToString(static_cast<OverlayHelperError>(exitCode));
-	qWarning("OverlayPrivateWin: overlay helper process exited (%s) with status code %s.",
+	qWarning("OverlayPrivateWin: overlay helper process '%s' exited (%s) with status code %s.",
+	         qPrintable(path),
 	         exitStatusString(exitStatus),
 	         helperErrString ? helperErrString : qPrintable(QString::number(exitCode)));
 
 	// If the helper process exited while we're in 'active'
 	// mode, restart it.
 	if (m_active) {
+		// If the helper was only recently started, be
+		// a little more patient with restarting it.
+		// We could be hitting a crash bug in the helper,
+		// and we don't want to do too much harm in that
+		// case by spawning thousands of processes.
+		qint64 cooldownMsec = (qint64) g.s.iOverlayWinHelperRestartCooldownMsec;
+		int delayMsec = g.s.iOverlayWinHelperRestartDelayMsec;
+		if (elapsedMsec < cooldownMsec) {
+			qWarning("OverlayPrivateWin: waiting up to %llu seconds until restarting helper process. last restart was %llu seconds ago.",
+				(unsigned long long) delayMsec/1000ULL,
+				(unsigned long long) elapsedMsec/1000ULL);
+			if (!m_delayedRestartTimer->isActive()) {
+				m_delayedRestartTimer->start(delayMsec);
```

Comment:
Why not elapsedMsec - cooldownMsec? If the goal is a rate limit of one restart per 10 seconds I'm not sure what iOverlayWinHelperRestartDelayMs does vs the cooldown.